tree/treeplayer/test/gh16804.cxx

```diff
@@ -0,0 +1,162 @@
+#include <cstdio>
+#include <fstream>
+#include <numeric>
+
+#include <TChain.h>
+#include <TFile.h>
+#include <TSystem.h>
+#include <TTree.h>
+#include <TTreePlayer.h>
+#include <TEntryList.h>
+#include <TDirectory.h>
+
+#include "gtest/gtest.h"
+
+// Test regression for https://github.com/root-project/root/issues/16804
+struct RegresssionGH16804 : public ::testing::Test {
+   constexpr static inline std::array<const char *, 2> fFriendFileNames{"gh16804_friend_0.root",
+                                                                        "gh16804_friend_1.root"};
+   constexpr static inline const char *fMainFileName{"gh16804_main.root"};
+   constexpr static inline const char *fMainTreeName{"mainTree"};
+   constexpr static inline const char *fFriendTreeName{"friendTree"};
+
+   static void CreateMainTree()
+   {
+      int begin{};
+      int end{20};
+
+      auto file = std::make_unique<TFile>(fMainFileName, "RECREATE");
+      auto tree = std::make_unique<TTree>(fMainTreeName, fMainTreeName);
+
+      int mainBranch{};
+      tree->Branch("mainBranch", &mainBranch);
+
+      // Sequential entries in reverse order from 19 (inclusive) to 0 (inclusive)
+      for (mainBranch = end - 1; mainBranch > begin - 1; mainBranch--)
+         tree->Fill();
+
+      file->Write();
+   }
+
+   static void CreateFriendTrees()
+   {
+      int begin{};
+      int end{10};
+      for (const auto &fn : fFriendFileNames) {
+         auto file = std::make_unique<TFile>(fn, "RECREATE");
+         auto tree = std::make_unique<TTree>(fFriendTreeName, fFriendTreeName);
+
+         int index{};
+         int value{100 + begin};
+         tree->Branch("index", &index);
+         tree->Branch("value", &value);
+
+         for (index = begin; index < end; ++index) {
+            tree->Fill();
+            value++;
+         }
+
+         file->Write();
+         begin += 10;
+         end += 10;
+      }
+   }
+
+   static void SetUpTestSuite()
+   {
+      CreateMainTree();
+      CreateFriendTrees();
+   }
+
+   static void TearDownTestSuite()
+   {
+      for (const auto &fn : fFriendFileNames)
+         std::remove(fn);
+
+      std::remove(fMainFileName);
+   }
+};
+
+TEST_F(RegresssionGH16804, MainTTreeFriendTChain)
+{
+   TChain friendChain{fFriendTreeName};
+   for (const auto &fn : fFriendFileNames)
+      friendChain.Add(fn);
+
+   auto mainFile = std::make_unique<TFile>(fMainFileName);
+   auto mainTree = mainFile->Get<TTree>(fMainTreeName);
+   mainTree->AddFriend(&friendChain);
```

Comment:
We should also test here-ish the case of a friend chain of a chain (in particular to double check we did not break it)